std_execution.bs

```diff
@@ -1243,6 +1244,133 @@ talk about the semantics of some of the algorithms proposed in this paper.
+semantics are able to make certain optimizations that are not available if
+senders can be potentially eager. With lazy senders, an algorithm can safely
+assume that a call to `execution::start` on an operation state strictly happens
+before any receiver methods are called. This frees the algorithm from needing to
+resolve potential race conditions between the start and completion of the work.
+
+When an algorithm needs to deal with potentially eager senders, the potential
+race conditions can be resolved one of several ways, none of which is desirable:
+
+1.  Assume the worst and implement the algorithm defensively, assuming all
+    senders are eager. This obviously has overheads both at runtime and in
+    algorithm complexity. Resolving race conditions is hard.
+
+2.  Require senders to declare whether they are eager or not with a query.
+    Algorithms can then implement two different implementation strategies, one
+    for strictly lazy senders and one for potentially eager senders. This
+    addresses the performance problem of (1) while compounding the complexity
+    problem.
+
+3.  Push the problem back onto eager senders by requiring them to never call a
+    receiver's `set_*` completion until after `execution::start` has been called
+    on the operation state. That puts the onus of dealing with the race
+    condition on the sender author, which is arguably where it belongs. But it
+    has the negative effect of delaying the completion of the operation. Even
+    when the work is done and the continuation available, the sender has to
+    withold the completion, possibly indefinitely.
+
+Of these options, (3) seems the least bad but would seem to run contrary to the
+<i>raison d'être</i> of eager algorithms, which ostensibly is to make the
+results available sooner.
+
+### Eager senders incur cancellation-related overhead ### {#design-lazy-algorithms-runtime}
+
+Another implication of the use of eager operations is with regards to
+cancellation. The eagerly executing operation will not have access to the
+caller's stop token until the sender is connected to a receiver. If we still
+want to be able to cancel the eager operation then it will need to create a new
+stop source and pass its associated stop token down to child operations. Then
+when the returned sender is eventually connected it will register a stop
+callback with the receiver's stop token that will request stop on the eager
+sender's stop source.
+
+As the eager operation does not know at the time that it is launched what the
+type of the receiver is going to be, and thus whether or not the stop token
+returned from `execution::get_stop_token` is an `std::unstoppable_token` or not,
+the eager operation is going to need to assume it might be later connected to a
+receiver with a stop token that might actually issue a stop request. Thus it
+needs to incur the runtime overhead of supporting cancellation, even if
+cancellation will never be requested by the caller.
```

Comment:
part of the cost of this is that the stop_token_callback must be type-erased with potential allocation.

Reply:
The type erasure is required (and indeed is implicit in the design of stop-tokens/stop-callbacks).
However, you shouldn't need a separate heap allocate storage for the stop-token callback - that should be stored inside the operation-state returned from connecting the eager sender.

Reply:
I don't understand. If you don't know the type of the receiver, you don't know the type of the stop callback for that receiver. It needs to be type erased, and since you don't know how big it will be, that could incur a dynamic allocation. What am I missing?